# **Filtering Crosslink-Spectrum-Matches and Crosslinks**

In [1]:
from pyXLMS import __version__

print(f"Installed pyXLMS version: {__version__}")

Installed pyXLMS version: 1.4.3


In [2]:
from pyXLMS import parser
from pyXLMS import transform

All data transformation functionality - including all filters - is available via the `transform` submodule. We also import the `parser` submodule here for reading result files.

In [3]:
parser_result = parser.read(
    "../../data/ms_annika/XLpeplib_Beveridge_QEx-HFX_DSS_R1.pdResult",
    engine="MS Annika",
    crosslinker="DSS",
)

Reading MS Annika crosslinks...: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 300/300 [00:00<00:00, 19994.78it/s]


We read crosslink-spectrum-matches and crosslinks using the [generic parser](https://hgb-bin-proteomics.github.io/pyXLMS/pyXLMS.parser.html#pyXLMS.parser.read) from a single `.pdResult` file.

In [4]:
csms = parser_result["crosslink-spectrum-matches"]
xls = parser_result["crosslinks"]

For easier access we assign our crosslink-spectrum-matches to the variable `csms` and our crosslinks to the variable `xls`.

## **Filtering by Crosslink Type**

In [5]:
csms_by_crosslink_type = transform.filter_crosslink_type(csms)

We can filter crosslink-spectrum-matches by their crosslink type (intra-links and inter-links) by calling `transform.filter_crosslink_type()` and passing the crosslink-spectrum-matches as the first argument. The function returns a dictionary containing the keys `"Intra"` and `"Inter"` with their associated values being lists of intra-link CSMs and inter-link CSMs respectively. You can read more about the `filter_crosslink_type()` function and all its parameters here: [**docs**](https://hgb-bin-proteomics.github.io/pyXLMS/pyXLMS.transform.html#pyXLMS.transform.filter.filter_crosslink_type).

In [6]:
print(f"Number of intra-link CSMs: {len(csms_by_crosslink_type['Intra'])}")
print(f"Number of inter-link CSMs: {len(csms_by_crosslink_type['Inter'])}")

Number of intra-link CSMs: 803
Number of inter-link CSMs: 23


Our example file contains 803 intra-link CSMs and 23 inter-link CSMs.

**Please note that any CSMs without associated protein accessions would be considered inter-links!**

In [7]:
xls_by_crosslink_type = transform.filter_crosslink_type(xls)

Similarly, we can filter crosslinks by their crosslink type (intra-links and inter-links) by calling `transform.filter_crosslink_type()` and passing the crosslinks as the first argument. The function returns a dictionary containing the keys `"Intra"` and `"Inter"` with their associated values being lists of intra-link crosslinks and inter-link crosslinks respectively. You can read more about the `filter_crosslink_type()` function and all its parameters here: [**docs**](https://hgb-bin-proteomics.github.io/pyXLMS/pyXLMS.transform.html#pyXLMS.transform.filter.filter_crosslink_type).

In [8]:
print(f"Number of intra-link XLs: {len(xls_by_crosslink_type['Intra'])}")
print(f"Number of inter-link XLs: {len(xls_by_crosslink_type['Inter'])}")

Number of intra-link XLs: 279
Number of inter-link XLs: 21


Our example file contains 279 intra-link crosslinks and 21 inter-link crosslinks.

**Please note that any crosslinks without associated protein accessions would be considered inter-links!**

## **Filtering by Peptide Pair**

In [9]:
csms_by_peptide_pair = transform.filter_peptide_pair_distribution(csms)

We can filter crosslink-spectrum-matches by their associated peptide pairs by calling `transform.filter_peptide_pair_distribution()` and passing the crosslink-spectrum-matches as the first argument. The function returns a dictionary containing the peptide pairs as keys with their associated values being lists of the corresponding CSMs. You can read more about the `filter_peptide_pair_distribution()` function and all its parameters here: [**docs**](https://hgb-bin-proteomics.github.io/pyXLMS/pyXLMS.transform.html#pyXLMS.transform.filter.filter_peptide_pair_distribution).

In [10]:
list(csms_by_peptide_pair.keys())[:5]

['GQKNSR-GQKNSR',
 'GQKNSR-GSQKDR',
 'SDKNR-SDKNR',
 'DKQSGK-DKQSGK',
 'DKQSGK-HSIKK']

Here are the first five peptide pairs that were encountered in our crosslink-spectrum-matches.

In [11]:
len(csms_by_peptide_pair["LSKSR-MKNYWR"])

9

For the peptide pair `LSKSR-MKNYWR` we found 9 crosslink-spectrum-matches in our result.

## **Filtering by Protein**

In [12]:
csms_by_protein = transform.filter_protein_distribution(csms)

We can filter crosslink-spectrum-matches by their associated proteins by calling `transform.filter_protein_distribution()` and passing the crosslink-spectrum-matches as the first argument. The function returns a dictionary containing the proteins as keys with their associated values being lists of the corresponding CSMs. You can read more about the `filter_protein_distribution()` function and all its parameters here: [**docs**](https://hgb-bin-proteomics.github.io/pyXLMS/pyXLMS.transform.html#pyXLMS.transform.filter.filter_protein_distribution).

In [13]:
list(csms_by_protein.keys())

['Cas9', 'sp']

In our example we have crosslink-spectrum-matches from two different proteins (or rather one protein `"Cas9"` and one protein group `"sp"` which denotes contaminants).

In [14]:
len(csms_by_protein["Cas9"])

821

In total we have 821 crosslink-spectrum-matches where at least one of the two crosslinked peptides is from Cas9.

In [15]:
xls_by_protein = transform.filter_protein_distribution(xls)

Similarly, we can filter crosslinks by their associated proteins by calling `transform.filter_protein_distribution()` and passing the crosslinks as the first argument. The function returns a dictionary containing the proteins as keys with their associated values being lists of the corresponding crosslinks. You can read more about the `filter_protein_distribution()` function and all its parameters here: [**docs**](https://hgb-bin-proteomics.github.io/pyXLMS/pyXLMS.transform.html#pyXLMS.transform.filter.filter_protein_distribution).

In [16]:
list(xls_by_protein.keys())

['Cas9', 'sp']

We also have crosslinks from two different proteins (or rather one protein `"Cas9"` and one protein group `"sp"` which denotes contaminants).

In [17]:
len(xls_by_protein["Cas9"])

295

In total we have 295 crosslinks where at least one of the two crosslinked peptides is from Cas9.

## **Getting Crosslink-Spectrum-Matches or Crosslinks of Specific Proteins**

In [18]:
csms_cas9 = transform.filter_proteins(csms, proteins=["Cas9"])

If we are only interested in crosslink-spectrum-matches of a specific protein (or set of proteins) we can further investigate this with the `transform.filter_proteins()` function and passing the crosslink-spectrum-matches as the first argument. The second argument `proteins` should be a list or set of protein accessions that we are interested in - in the example here we are only interested in a single protein, namely `"Cas9"`.

In [19]:
list(csms_cas9.keys())

['Proteins', 'Both', 'One']

The function returns a dictionary with keys `"Proteins"`, `"Both"`, and `"One"`:
- `"Proteins"` allows you to access your original list of proteins that was used for filtering (e.g., what was passed via the `proteins` parameter).
- `"Both"` contains all crosslink-spectrum-matches where both peptides were of one of the specified proteins, in our case both peptides are from Cas9.
- `"One"` contains all crosslink-spectrum-matches where one of the two crosslinked peptides was of the specified proteins, in our case from Cas9.

In [30]:
csms_cas9["Proteins"]

['Cas9']

Via `"Proteins"` we can access our original list of proteins that was used for filtering.

In [32]:
len(csms_cas9["Both"])

798

Via `"Both"` we get all crosslink-spectrum-matches where both peptides are of one of the specified proteins of interest (in our case there was only one protein of interest: Cas9).

In [33]:
len(csms_cas9["One"])

23

In [34]:
xls_cas9 = transform.filter_proteins(xls, proteins=["Cas9"])

In [35]:
xls_cas9["Proteins"]

['Cas9']

In [37]:
len(xls_cas9["Both"])

274

In [38]:
len(xls_cas9["One"])

21

## **Filtering by Target-Decoy Type**

In [39]:
csms_td = transform.filter_target_decoy(csms)

In [40]:
len(csms_td["Target-Target"])

786

In [41]:
len(csms_td["Target-Decoy"])

39

In [42]:
len(csms_td["Decoy-Decoy"])

1

In [43]:
xls_td = transform.filter_target_decoy(xls)

In [44]:
len(xls_td["Target-Target"])

265

In [45]:
len(xls_td["Target-Decoy"])

0

In [46]:
len(xls_td["Decoy-Decoy"])

35

## **Filtering Target Matches Only**

In [47]:
csms_tt = transform.targets_only(csms)

In [48]:
len(csms_tt)

786

In [49]:
xls_tt = transform.targets_only(xls)

In [50]:
len(xls_tt)

265

In [51]:
parser_result = transform.targets_only(parser_result)

In [52]:
len(parser_result["crosslink-spectrum-matches"])

786

In [53]:
len(parser_result["crosslinks"])

265